core/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java

```diff
@@ -0,0 +1,1001 @@
+
+    /** Resets the call counters for all the trackers in the map. */
+    void resetCallCounters() {
+      for (AddressTracker tracker : trackerMap.values()) {
+        tracker.clearCallCounters();
+      }
+    }
+
+    /**
+     * When OD gets disabled we need to uneject any subchannels that may have been ejected and
+     * to reset the ejection time multiplier.
+     */
+    void cancelTracking() {
+      for (AddressTracker tracker : trackerMap.values()) {
+        if (tracker.subchannelsEjected()) {
+          tracker.unejectSubchannels();
+        }
+        tracker.resetEjectionTimeMultiplier();
+      }
+    }
+
+    /** Swaps the active and inactive counters for each tracker. */
+    void swapCounters() {
+      for (AddressTracker tracker : trackerMap.values()) {
+        tracker.swapCounters();
+      }
+    }
+
+    /**
+     * At the end of a timer run we need to decrement the ejection time multiplier for trackers
+     * that don't have ejected subchannels and uneject ones that have spent the maximum ejection
+     * time allowed.
+     */
+    void maybeUnejectOutliers(Long detectionTimerStartNanos) {
+      for (AddressTracker tracker : trackerMap.values()) {
+        if (!tracker.subchannelsEjected()) {
+          tracker.decrementEjectionTimeMultiplier();
+        }
+
+        if (tracker.subchannelsEjected() && tracker.maxEjectionTimeElapsed(
+            detectionTimerStartNanos)) {
+          tracker.unejectSubchannels();
+        }
+      }
+    }
+
+    /** Returns only the trackers that have the minimum configured volume to be considered. */
+    List<AddressTracker> trackersWithVolume(OutlierDetectionLoadBalancerConfig config) {
+      List<AddressTracker> trackersWithVolume = new ArrayList<>();
+      for (AddressTracker tracker : trackerMap.values()) {
+        if (tracker.inactiveVolume() >= config.successRateEjection.requestVolume) {
+          trackersWithVolume.add(tracker);
+        }
+      }
+      return trackersWithVolume;
+    }
+
+    /**
+     * How many percent of the addresses would have their subchannels ejected if we proceeded
+     * with the next ejection.
+     */
+    double nextEjectionPercentage() {
+      int totalAddresses = 0;
+      int ejectedAddresses = 0;
+      for (AddressTracker tracker : trackerMap.values()) {
+        totalAddresses++;
+        if (tracker.subchannelsEjected()) {
+          ejectedAddresses++;
+        }
+      }
+      return ((double)(ejectedAddresses + 1) / totalAddresses) * 100;
+    }
+  }
+
+
+  /**
+   * Implementations provide different ways of ejecting outlier addresses..
+   */
+  interface OutlierEjectionAlgorithm {
+
+    /**
+     * Is the given {@link EquivalentAddressGroup} an outlier based on the past call results stored
+     * in {@link AddressTracker}.
+     */
+    void ejectOutliers(AddressTrackerMap trackerMap, long ejectionTimeMillis);
+
+    @Nullable
+    static OutlierEjectionAlgorithm forConfig(OutlierDetectionLoadBalancerConfig config) {
```

Comment:
This function produces a single algorithm object for a given config, but a config is allowed to enable both algorithms. You may need to change other code to account for having multiple active algorithms.

Reply:
Ah, I had assumed it was an either/or situation. Updated now to support an arbitrary amount of ejection algorithms.